File: kaalkaNTP/packet.py
# kaalka/packet.py
import socket
from kaalkaNTP import KaalkaNTP

class Packet:
    def __init__(self, data: str) -> None:
        self.data = data
        self.encrypted_data = None

    def encrypt(self, kaalka: KaalkaNTP) -> None:
        if not self.encrypted_data:
            encrypted_message = kaalka.encrypt(self.data)
            self.encrypted_data = encrypted_message

    def decrypt(self, kaalka: KaalkaNTP) -> None:
        if self.encrypted_data:
            decrypted_message = kaalka.decrypt(self.encrypted_data)
            self.data = decrypted_message
            self.encrypted_data = None

    @staticmethod
    def sender():
        # Simulate sender preparing data
        message = "Hello, Kaalka!"
        print("Original Message:", message)

        # Create a packet with the message
        packet = Packet(message)

        # Encrypt the packet using Kaalka algorithm
        kaalka = KaalkaNTP()
        packet.encrypt(kaalka)
        print("Encrypted Data:", packet.encrypted_data)

        # Simulate sending the encrypted data over the network
        Packet.send_data_over_network(packet.encrypted_data)

    @staticmethod
    def receiver():
        # Simulate receiving the encrypted data over the network
        received_data = Packet.receive_data_over_network()

        # Create a packet with the received data
        packet = Packet(received_data)

        # Decrypt the packet using Kaalka algorithm
        kaalka = KaalkaNTP()
        packet.decrypt(kaalka)
        print("Decrypted Message:", packet.data)

    @staticmethod
    def send_data_over_network(data):
        # Simulate sending data over the network (e.g., using sockets)
        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
            s.sendto(data.encode(), ('127.0.0.1', 12345))

    @staticmethod
    def receive_data_over_network():
        # Simulate receiving data over the network (e.g., using sockets)
        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
            s.bind(('127.0.0.1', 12345))
            data, addr = s.recvfrom(1024)
            return data.decode()

File: packet.js
const dgram = require('dgram');
const KaalkaNTP = require('./kaalkaNTP');

class Packet {
  constructor(data) {
    this.data = data;
    this.encryptedData = null;
  }

  async encrypt(kaalka) {
    if (!this.encryptedData) {
      const encryptedMessage = await kaalka.encrypt(this.data);
      this.encryptedData = encryptedMessage;
    }
  }

  async decrypt(kaalka) {
    if (this.encryptedData) {
      const decryptedMessage = await kaalka.decrypt(this.encryptedData);
      this.data = decryptedMessage;
      this.encryptedData = null;
    }
  }

  static async sender() {
    // Simulate sender preparing data
    const message = "Hello, Kaalka!";
    console.log("Original Message:", message);

    // Create a packet with the message
    const packet = new Packet(message);

    // Encrypt the packet using Kaalka algorithm
    const kaalka = new KaalkaNTP();
    await packet.encrypt(kaalka);
    console.log("Encrypted Data:", packet.encryptedData);

    // Simulate sending the encrypted data over the network
    await Packet.sendDataOverNetwork(packet.encryptedData);
  }

  static async receiver() {
    // Simulate receiving the encrypted data over the network
    const receivedData = await Packet.receiveDataOverNetwork();

    // Create a packet with the received data
    const packet = new Packet(receivedData);

    // Decrypt the packet using Kaalka algorithm
    const kaalka = new KaalkaNTP();
    await packet.decrypt(kaalka);
    console.log("Decrypted Message:", packet.data);
  }

  static async sendDataOverNetwork(data) {
    // Simulate sending data over the network (e.g., using dgram sockets)
    const socket = dgram.createSocket('udp4');
    return new Promise((resolve) => {
      socket.send(data, 0, data.length, 12345, '127.0.0.1', (err) => {
        socket.close();
        resolve();
      });
    });
  }

  static async receiveDataOverNetwork() {
    // Simulate receiving data over the network (e.g., using dgram sockets)
    const socket = dgram.createSocket('udp4');
    return new Promise((resolve) => {
      socket.on('message', (msg) => {
        socket.close();
        resolve(msg.toString());
      });
      socket.bind(12345, '127.0.0.1');
    });
  }
}

module.exports = Packet;
